Fix up chunked uploading

xloem opened the pull request:
Somehow in transitions it looks like the chunked uploading feature broke.

This change fixes it up again.

Chunked uploading is done by passing an iterator as the body data; see https://requests.readthedocs.io/en/master/user/advanced/#chunk-encoded-requests

I believe you can also pass a file-like object to stream with known size for #30

m-cat reviewed (changes requested):
Thanks a lot for the contribution. Couple comments:

1. Can you please create a quick test for this function? See the existing upload tests for examples.
1. Lint CI is failing

m-cat commented:
Hmm should we keep the `path` parameter? If `custom_filename` is not passed then the filename is empty. It would be consistent with the Go SDK where we require a filename when uploading generic data.

xloem commented:
The existing tests are uploading a 5B file and verifying the upload matches the skylink for a 156KB pdf ... this does not seem correct ...

xloem commented:
Whew!  Confused myself with tests

> Hmm should we keep the `path` parameter? If `custom_filename` is not passed then the filename is empty. It would be consistent with the Go SDK where we require a filename when uploading generic data.

I've made the interface change you requested, but I'm noting that this changes the interface from how chunked uploading used to be here.

I like the way the go library takes a map of filenames to content: do you think that would work here?

m-cat commented:
> I like the way the go library takes a map of filenames to content: do you think that would work here?

That would be excellent! Ideally the SDKs are consistent with each other, especially the APIs, but I am short on time. So these kinds of fixes are really appreciated!

xloem commented:
Looking at this some, I'm noticing that chunked uploading (which wonderfully supports live streaming uploads) works for single file uploads only with the requests library.  Additionally the go SDK does not have a feature for chunked uploads.  It seems like supporting generic upload, which the go SDK has normalized, is a separate beast from supporting chunked upload.  It might not be hard to upload an iterator as chunked if only one file is provided.

xloem commented:
@m-cat I've squashed and rebased and simplified to addition of a method comparable to the generic one used by the go sdk.  It takes a map of filenames to file-like objects or bytes data, and if there is only one file can do chunked uploading via the requests library's feature, if the file data is an iterator.  Does this look good?

m-cat commented on siaskynet/_upload.py:
```diff
@@ -18,6 +18,67 @@ def default_upload_options():
+
+    if len(upload_data) == 1:
+        fieldname = opts['portal_file_fieldname']
+        # this appears to be missing in go sdk; maybe the server ignores it?
```
Please write an issue on the Go SDK instead of a comment here, as this will go forgotten.

xloem replied:
I've asked about this at https://github.com/NebulousLabs/go-skynet/issues/26

m-cat replied:
Awesome, thanks 🙂

m-cat replied:
Can you also remove this comment?

xloem replied:
yes

m-cat commented:
Sorry for the delay in reviewing, we were super busy. Taking a look now.

m-cat commented on siaskynet/_upload.py:
```diff
@@ -48,41 +109,12 @@ def upload_file_request(self, path, custom_opts=None):
+        upload_data = {opts['custom_filename']: file_h}
 
-
-def upload_file_request_with_chunks(self, path, custom_opts=None):
```
We can't delete any methods because it would break backwards compatibility. I would keep this and just have it call the new method

xloem replied:
i can do this but i think there are compatibility issues elsewhere too; i originally made this PR while porting my code to the newer interfaces.

m-cat commented on siaskynet/__init__.py:
```diff
@@ -24,7 +24,8 @@ class SkynetClient():
         get_skykeys
     )
     from ._upload import (
-        upload_file, upload_file_request, upload_file_request_with_chunks,
```
Please export `upload_file_request_with_chunks` as starting in `v2` we are guaranteeing backwards compatibility for the SDKs.

xloem replied:
exported

m-cat commented on siaskynet/_upload.py:
```diff
@@ -48,41 +109,12 @@ def upload_file_request(self, path, custom_opts=None):
-            else os.path.basename(file_h.name)
-        files = {opts['portal_file_fieldname']: (filename, file_h)}
+        if not opts['custom_filename']:
+            opts['custom_filename'] = os.path.basename(file_h.name)
```
I would use a temporary variable here instead of overriding the values in `opts`. See the Go SDK

m-cat replied:
Oh, I see why you do this. The `filename` parameter is only needed for uploading directories, so this comment is correct:

```py
        # this appears to be missing in go sdk; maybe the server ignores it?
```

m-cat replied:
And if someone passes a map of filenames to files (e.g. `upload_data`) then we can assume the filename there is the one he wants to use. `custom_filename` should only apply to `upload_file` and not to `upload`, IMO.

xloem replied:
Note that `opts` is a copy, so overriding is invisible to the user.  I'll use a temporary variable like in the go sdk, though.

m-cat commented on siaskynet/_upload.py:
```diff
@@ -18,6 +18,67 @@ def default_upload_options():
+                not isinstance(data, bytes) and
+                not isinstance(data, str) and
+                not hasattr(data, 'read')):
+            # an iterator for chunked uploading
```
This looks good but it should be documented so people know this functionality is there

xloem replied:
is the documentation I added to `upload_file_request` sufficient?

m-cat reviewed (changes requested):
Looks good apart from a few points. Thanks for taking the time to make this conform to the Go SDK. BTW, have you tested uploads to skynet with this code?

m-cat commented:
Just ran into an issue with the Go SDK that I think applies here too. If we upload a directory with a single file, it will be treated as a file upload instead of a directory upload. This can be unexpected for the user, and a directory should be uploaded even if it's only a single file.

xloem commented:
@m-cat I believe I've added all the changes you requested.

I'm testing this locally and I discovered that python requests doesn't stream file objects when they are passed as multipart data (it reads the whole content of each file in one go).  If you pass a single file as the body of a post request, it _does_ stream it, reading it in a configurable chunk size defaulting to 8K.

Would it be reasonable to change the behavior of python-skynet to pass all single files directly as body content?  This would simplify the implementation and provide streaming for all single files.  The tests would change to expect this.

m-cat commented on siaskynet/_upload.py:
```diff
@@ -47,17 +48,19 @@ def upload_request(self, upload_data, custom_opts=None):
                         (filename, data)))
 
-    if len(upload_data) == 1:
+    if opts['portal_file_fieldname'] == fieldname:
```
This isn't very elegant. I would change this check

xloem replied:
I've changed it.  The reason the code is awkward here is I wanted to keep it similar to the go code to ease new changes.  Let me know if you have more preferences around it.

m-cat reviewed (approved):
Looks great! Thanks for your patience! Approving this on its 1-month anniversary 🙂

m-cat commented:
Published version 2.1.0 which includes these changes 👍
